src/org/zaproxy/zap/extension/websocket/WebSocketAPI.java

```diff
@@ -0,0 +1,444 @@
+        if (sent) {
+            try {
+                extension.recordMessage(resp);
+            } catch (DatabaseException e) {
+                LOG.error(e.getMessage(), e);
+            }
+        }
+        return sent;
+    }
+
+    private Map<Integer, EventConsumer> evMap = new HashMap<Integer, EventConsumer>();
+
+    private EventConsumer getEventConsumer(int channelId) {
+        EventConsumer ev = evMap.get(channelId);
+        if (ev == null) {
+            ev = new WebsocketEventConsumer(channelId);
+            evMap.put(channelId, ev);
+        }
+        return ev;
+    }
+
+    @Override
+    public ApiResponse handleApiView(String name, JSONObject params) throws ApiException {
+        ApiResponse result;
+
+        if (VIEW_CHANNELS.equals(name)) {
+            ApiResponseList resultList = new ApiResponseList(name);
+            try {
+                List<WebSocketChannelDTO> channels = extension.getChannels(new WebSocketChannelDTO());
+                for (WebSocketChannelDTO channel : channels) {
+                    Map<String, String> map = new HashMap<String, String>();
+                    map.put("id", Integer.toString(channel.id));
+                    map.put("displayName", channel.toString());
+                    map.put("connected", Boolean.toString(channel.isConnected()));
+                    map.put("inScope", Boolean.toString(channel.isInScope()));
+                    map.put("contextUrl", channel.getContextUrl());
+                    map.put("fullUrl", channel.getFullUri());
+                    if (channel.getHandshakeReference() != null) {
+                        map.put("handshakeHistoryId", Integer.toString(channel.getHandshakeReference().getHistoryId()));
+                    }
+                    resultList.addItem(new ApiResponseSet<String>("channel", map));
+                }
+                result = resultList;
+            } catch (DatabaseException e) {
+                LOG.error(e.getMessage(), e);
+                throw new ApiException(ApiException.Type.INTERNAL_ERROR);
+            }
+        } else if (VIEW_MESSAGE.equals(name)) {
+            try {
+                WebSocketMessageDTO message = extension
+                        .getWebsocketMessage(params.getInt(PARAM_MESSAGE_ID), params.getInt(PARAM_CHANNEL_ID));
+                result = wsMessageToResult(message, true);
+            } catch (DatabaseException e) {
+                LOG.error(e.getMessage(), e);
+                throw new ApiException(ApiException.Type.INTERNAL_ERROR);
+            }
+        } else if (VIEW_MESSAGES.equals(name)) {
+            ApiResponseList resultList = new ApiResponseList(name);
+            try {
+                WebSocketMessageDTO criteria = new WebSocketMessageDTO();
+                List<Integer> opcodes = null;
+                List<Integer> inScopeChannelIds = null;
+                int offset = this.getParam(params, PARAM_START, 0);

```

Comment:
Validate that's positive? Getting `java.sql.SQLDataException: data exception: invalid row count in result offset clause` with negative value.